gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+    conns = np.array(unique_list(connections))
+    return conns
+
+
+def create_dummy_loader(
+    tensor_x_t: Tensor,
+) -> DataLoader[Tuple[Tensor, ...]]:
+    """
+    A helper file for topological regularizer
+    A function for formatting the data for evaluating
+    the linear surrogate function.
+    Args:
+        tensor_x_t: Torch tensor representing the nodes of
+        the linear surrogate function.
+    Returns
+    -------
+        dummy_loader : A torch dataloader based off of tensor_x_t
+    """
+    dummy = np.zeros(len(tensor_x_t))
+    dummy_ind = torch.from_numpy(dummy)
+    dummy_set = TensorDataset(tensor_x_t, dummy_ind)
+    dummy_loader = DataLoader(dummy_set, batch_size=4)
+    return dummy_loader
+
+
+def unique_list(a_list: list) -> list:
+    """
+    A helper file for topological regularizer
+    Given a list a_list,
+    returns the unique elements in that.
+    Used for computing the connections
+    when building the linear surrogate function
+    Args:
+         a_list: a list
+    Returns
+    -------
+        uniquelist : a list of unique entries of the list a_list
+    """
+    uniquelist = []
+    used = set()
+    for item in a_list:
+        tmp = repr(item)
+        if tmp not in used:
+            used.add(tmp)
+            uniquelist.append(item)
+    return uniquelist
+
+
+def evaluate_model_on_grid(
+    dummy_loader: DataLoader[Tuple[Tensor, ...]],
+    model: torch.nn.Module,
+) -> np.ndarray:
+    """
+    A helper file for topological regularizer
+    Evaluates a binary classifier neural netowrk
+    on a dataset and returns the class 1 probabilities. This is used for
+    computing the persistence of the linear surrogate function.
+    Args:
+         dummy_loader: Torch dataloader representing
+             the knots of the linear surrogate function.
+             The response variable can be anything.
+         model: Torch model
+    Returns
+    -------
+        results : Class 1 probabilities of the model evaluated on
+            the dummy_loader
+    """
+    model.eval()
+    featdim = next(iter(dummy_loader))[0].shape[1]
+    # featdim = dummy_loader.dataset.tensors[0].shape[1]
+    preds = []
+    for i, data in enumerate(dummy_loader, 0):
+        inputs, labels = data
+        inputs = torch.reshape(inputs, (-1, featdim)).float()
+        outputs = model(inputs)
+        preds.append(F.softmax(outputs, 1)[:, 1])
+    tmp = [p.detach().numpy() for p in preds]
+    results = np.concatenate(tmp, axis=0)
+    return results
+
+
+def get_persistent_pairs(
+    results: np.ndarray, connections: np.ndarray, expansion: int = 6
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    A helper file for topological regularizer
+    Given Class 1 probabilities and node connectivity information,
+    returns the persistent pairs of the function
+    associated to filtering according to the numerical values
+    Args:
+        results: an n-vector of class 1 probabilities
```

Comment:
same cimment as above on formatting. please skip a line and add 4 spaces

Reply:
done